src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -1340,6 +1341,43 @@ public bool WaitForExit(int milliseconds)
             return exited;
         }
 
+        /// <summary>
+        /// Instructs the Process component to wait for the associated process to exit, or
+        /// for the <paramref name="cancellationToken"/> to be canceled.
+        /// </summary>
+        public async Task WaitForExitAsync(CancellationToken cancellationToken = default)
+        {
+            if (!Associated) { throw new InvalidOperationException(SR.NoAssociatedProcess); }
+            if (!EnableRaisingEvents) { throw new InvalidOperationException(SR.EnableRaisingEventsRequired); }
+
+            var tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            void ExitedHandler(object sender, EventArgs e)
+            {
+                tcs.TrySetResult(null);
+            }
+
+            Exited += ExitedHandler;
+
+            try
+            {
+                if (HasExited)
+                {
+                    // Handle the race where the process exited prior to registering the event handler
+                    return;
+                }
+
+                using (cancellationToken.Register(() => tcs.TrySetCanceled()))
```

Comment:
Use the overload that lets you pass an object in here, to avoid allocating a delegate per call.

Reply:
It could also use UnsafeRegister instead of Register.  We don't need ExecutionContext to flow into the callback here, because there's no user code getting invoked. e.g.
```C#
using (cancellationToken.UnsafeRegister(s => ((TaskCompletionSource<object>)s).TrySetCanceled(), tcs))
```

Ideally, we'd also pass `cancellationToken` into `TrySetCanceled` as an argument, so that the thrown OperationCanceledException contains the relevant token.  Though to do that without incurring additional allocations would require creating a TCS-derived type that could store the CancellationToken on it as a field... and it's probably not worth it.  An alternate that's not quite as good but still relatively pay for play would be to change the await below to be:
```C#
try
{
    await tcs.Task.ConfigureAwait(false);
}
catch (OperationCanceledException)
{
    cancellationToken.ThrowIfCancellationRequested();
    throw;
}
```

Reply:
> Though to do that without incurring additional allocations would require creating a TCS-derived type that could store the CancellationToken on it as a field... and it's probably not worth it.

We could promote the `TaskCompletionSourceWithCancellation` class from System.Net.Http to common and reuse it here.

Reply:
I'm happy to use `TaskCompletionSourceWithCancellation`, and it'd be nice to do at least one thing "ideally" ;).

I've not introduced a common class before; from looking at the csproj, it looks like the process is to move the class into the appropriate namespace under common, then add the file as a link in both libraries? Please let me know if I'm missing anything!

Reply:
You'll want to move the file into the Common\src\System\Threading\Tasks directory, change namespace, make sure the original library still compiles.

Reply:
I've gone ahead and moved `TaskCompletionSourceWithCancellation` to common, wired it up to both projects, and verified locally that it still compiles and test pass in both.